janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java

```diff
@@ -778,27 +778,42 @@ private String getDeletionScript(KeyInformation.IndexRetriever information, Stri
     private String getAdditionScript(KeyInformation.IndexRetriever information, String storeName,
                                      IndexMutation mutation) throws PermanentBackendException {
         final StringBuilder script = new StringBuilder();
+
         for (final IndexEntry e : mutation.getAdditions()) {
             final KeyInformation keyInformation = information.get(storeName).get(e.field);
-            switch (keyInformation.getCardinality()) {
-                case SET:
-                case LIST:
-                    script.append("if(ctx._source[\"").append(e.field).append("\"] == null) ctx._source[\"").append(e.field).append("\"] = [];");
-                    script.append("ctx._source[\"").append(e.field).append("\"].add(").append(convertToJsType(e.value, compat.scriptLang(), Mapping.getMapping(keyInformation))).append(");");
-                    if (hasDualStringMapping(keyInformation)) {
-                        script.append("if(ctx._source[\"").append(getDualMappingName(e.field)).append("\"] == null) ctx._source[\"").append(getDualMappingName(e.field)).append("\"] = [];");
-                        script.append("ctx._source[\"").append(getDualMappingName(e.field)).append("\"].add(").append(convertToJsType(e.value, compat.scriptLang(), Mapping.getMapping(keyInformation))).append(");");
-                    }
-                    break;
-                default:
-                    break;
+            final Cardinality cardinality = keyInformation.getCardinality();
 
+            if (cardinality != Cardinality.SET && cardinality != Cardinality.LIST) {
+                continue;
             }
 
+            String value = convertToJsType(e.value, compat.scriptLang(), Mapping.getMapping(keyInformation));
+
+            appendAdditionScript(script, value, e.field, cardinality == Cardinality.SET);
+
+            if (hasDualStringMapping(keyInformation)) {
+                appendAdditionScript(script, value, getDualMappingName(e.field), cardinality == Cardinality.SET);
+            }
         }
+
         return script.toString();
     }
 
+    private void appendAdditionScript(StringBuilder script, String value, String fieldName, Boolean distinct) {
+        script.append("if (ctx._source[\"").append(fieldName).append("\"] == null) ctx._source[\"").append(fieldName).append("\"] = [];");
+
+        if (distinct) {
+            script.append("def index = ctx._source[\"").append(fieldName).append("\"].indexOf(").append(value).append(");");
```

Comment:
Maybe I am missing something but isn't `_source[fieldName]` returns a collection? If so, how can we use `indexOf` here? Maybe we should use `.contains(value)`?
P.S. Maybe I am missing something.

Reply:
AFAIK it returns ArrayList. Maybe it's some kind of hack and by contract it returns collection, but it works. Take a look at `getDeletionScript` that uses `indexOf` call to remove elements. I've just reused the same approach.

Reply:
My bad. I see ArrayList has `indexOf` method. Also I see that method `contains` actually uses `indexOf` in it's ArrayList implementation.